### PR TITLE
soc: arm: st_stm32: stm32l0: Add RTC support

### PR DESCRIPTION
**Pull request by baw-serafin**
Tested with `samples/drivers/counter/alarm/` on `b_l072z_lrwan1`

### CODE REVIEW

**Comment by baw-serafin** [comment]
Should I split the documentation into separate commit?

**Comment by baw-serafin** [comment]
> Should I split the documentation into separate commit?

Done by erwango's request

**Review by erwango — CHANGES_REQUESTED** [review]

File: boards/arm/b_l072z_lrwan1/b_l072z_lrwan1_defconfig [hunk 1]
```diff
@@ -21,6 +21,8 @@ CONFIG_CLOCK_STM32_PLL_DIVISOR=2
 CONFIG_CLOCK_STM32_PLL_MULTIPLIER=4
 CONFIG_CLOCK_STM32_AHB_PRESCALER=1
 CONFIG_CLOCK_STM32_APB1_PRESCALER=1
+# Use LSE source for RTC
+CONFIG_COUNTER_RTC_STM32_CLOCK_LSE=y
```
[the remark this answers is not in the document]
Comment (baw-serafin) [reply boards/arm/b_l072z_lrwan1/b_l072z_lrwan1_defconfig]: done in 74d292f

**Review by erwango — CHANGES_REQUESTED** [review]
Please rebase, following merge of #23285.

**Comment by baw-serafin** [comment]
> Please rebase, following merge of #23285.

Done

**Review by FRASTM — APPROVED** [review]

**Comment by baw-serafin** [comment]
rebased on current master to solve conflict

File: soc/arm/st_stm32/stm32l0/dts_fixup.h [hunk 2]
```diff
@@ -88,4 +88,11 @@
 
 #define DT_ADC_1_NAME			DT_ST_STM32_ADC_40012400_LABEL
 
+#define DT_RTC_0_BASE_ADDRESS           DT_ST_STM32_RTC_40002800_BASE_ADDRESS
+#define DT_RTC_0_IRQ_PRI                DT_ST_STM32_RTC_40002800_IRQ_0_PRIORITY
+#define DT_RTC_0_IRQ                    DT_ST_STM32_RTC_40002800_IRQ_0
+#define DT_RTC_0_NAME                   DT_ST_STM32_RTC_40002800_LABEL
+#define DT_RTC_0_CLOCK_BITS             DT_ST_STM32_RTC_40002800_CLOCK_BITS
+#define DT_RTC_0_CLOCK_BUS              DT_ST_STM32_RTC_40002800_CLOCK_BUS
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l0/dts_fixup.h]: These should not be needed anymore after rebase.

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32l0/dts_fixup.h (hunk 2 above)
Comment (baw-serafin) [reply soc/arm/st_stm32/stm32l0/dts_fixup.h]: They are for the sample application:
```
.../zephyr/samples/drivers/counter/alarm/src/main.c: In function 'main':
.../zephyr/samples/drivers/counter/alarm/src/main.c:61:35: error: 'DT_RTC_0_NAME' undeclared (first use in this function); did you mean 'DT_I2C_2_NAME'?
   61 |  counter_dev = device_get_binding(DT_RTC_0_NAME);
      |                                   ^~~~~~~~~~~~~
      |                                   DT_I2C_2_NAME
.../zephyr/samples/drivers/counter/alarm/src/main.c:61:35: note: each undeclared identifier is reported only once for each function it appears in
[34/113] Building C object zephyr/CMakeFiles/zephyr.dir/drivers/timer/cortex_m_systick.c.obj
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: cmake --build /..../zephyr/build
serafin@sl-baw:../zephyr$ cd build/
serafin@sl-baw:../zephyr/build$ grep -nrH RTC_0
zephyr/b_l072z_lrwan1.dts.pre.tmp:49:   label = "RTC_0";
zephyr/include/generated/devicetree.conf:703:DT_ST_STM32_RTC_40002800_LABEL="RTC_0"
zephyr/include/generated/devicetree.conf:704:DT_INST_0_ST_STM32_RTC_LABEL="RTC_0"
zephyr/include/generated/devicetree_unfixed.h:772:#define DT_ST_STM32_RTC_40002800_LABEL              "RTC_0"
zephyr/zephyr.dts:47:			label = "RTC_0";
```

Comment (erwango) [reply soc/arm/st_stm32/stm32l0/dts_fixup.h]: My bad, I was not precise enough: Please remove all except `DT_RTC_0_NAME`.

File: boards/arm/b_l072z_lrwan1/Kconfig.defconfig [hunk 3]
```diff
@@ -28,4 +28,11 @@ config SPI_2
 	default y
 endif # SPI
 
+if COUNTER
+choice COUNTER_RTC_STM32_CLOCK_SRC
+	default COUNTER_RTC_STM32_CLOCK_LSE
+	depends on COUNTER_RTC_STM32
```
Comment (erwango) [inline boards/arm/b_l072z_lrwan1/Kconfig.defconfig:34]: Please remove the "if COUNTER" and replace this line with "depends on COUNTER"

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**